src/server/io_utils.cc

```diff
@@ -0,0 +1,62 @@
+// Copyright 2022, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/io_utils.h"
+
+#include "base/flags.h"
+#include "base/logging.h"
+#include "server/error.h"
+
+using namespace std;
+
+namespace dfly {
+
+io::Result<size_t> BufferedStreamerBase::WriteSome(const iovec* vec, uint32_t len) {
+  // Write data to producer buffer.
+  buffered_++;
+  return io::BufSink{&producer_buf_}.WriteSome(vec, len);
+}
+
+void BufferedStreamerBase::NotifyWritten() {
+  // Wake up the consumer.
+  waker_.notify();
+  // Block if we're stalled because the consumer is not keeping up.
+  waker_.await([this]() { return !IsStalled() || IsStopped(); });
+}
+
+error_code BufferedStreamerBase::ConsumeIntoSink(io::Sink* dest) {
+  while (!IsStopped()) {
+    // Wait for more data or stop signal.
+    waker_.await([this]() { return buffered_ > 0 || IsStopped(); });
+    if (IsStopped())
```

Comment:
If producer is done doesnt it make more sense to finish write all the buffered journal data and then return?

Reply:
Indeed.
It doesn't matter currently because its used as cancellation, but the general behavior should be this